Fix embedding example, simplify Makefile, and test it

ihnorton opened the pull request:
- get rid of makefile cruft
- build in-place instead of putting in usr/bin
- use julia-config for flags, as documented in manual

(related: #18421)

ihnorton committed:
Fix embedding example and simplify Makefile

- get rid of makefile cruft
- use julia-config for flags, as documented in manual
- build into `libexec`

also move to subdir for clarity.

tkelman commented on examples/Makefile:
```diff
@@ -1,41 +1,25 @@
-JULIAHOME = $(abspath ..)
-include $(JULIAHOME)/Make.inc
+JULIA = $(abspath ../usr/bin/julia)
```
this isn't right for an out-of-tree build

tkelman commented on examples/Makefile:
```diff
@@ -1,41 +1,25 @@
-JULIAHOME = $(abspath ..)
-include $(JULIAHOME)/Make.inc
+JULIA = $(abspath ../usr/bin/julia)
+JL_SHARE = $(shell $(JULIA) -e 'print(joinpath(JULIA_HOME,Base.DATAROOTDIR,"julia"))')
```
may require spawn

tkelman commented on examples/Makefile:
```diff
@@ -1,41 +1,25 @@
-ifneq ($(UNTRUSTED_SYSTEM_LIBM),0)
-JLDFLAGS += $(WHOLE_ARCHIVE) $(build_libdir)/libopenlibm.a $(NO_WHOLE_ARCHIVE)
-endif
+ifeq ($(OS),Windows_NT)
```
not correct for cross-compiling

tkelman commented on src/julia.h:
```diff
@@ -1313,6 +1313,7 @@ JL_DLLEXPORT void julia_init(JL_IMAGE_SEARCH rel);
 JL_DLLEXPORT void jl_init(const char *julia_home_dir);
 JL_DLLEXPORT void jl_init_with_image(const char *julia_home_dir,
                                      const char *image_relative_path);
+JL_DLLEXPORT const char* jl_get_default_sysimg_path();
```
if exported, needs JL_DLLEXPORT on both prototype and implementation

ihnorton commented:
Updated, thanks. I'm going for non-reliance on `Make.inc`, so that this can serve as a usable template. However, added checks for spawn and BUILDROOT independently (to handle cross-compile and out-of-tree). Hopefully sufficient, though I'm not set up to test those right now.

I'll note that this example isn't built at all right now with out-of-tree.

ihnorton commented:
Win64 and OSX test failures seem unrelated.

ihnorton committed:
Fix 20580, embedding example crash

tkelman commented on examples/Makefile:
```diff
@@ -1,41 +0,0 @@
-
-ifeq ($(USE_SYSTEM_LIBM),0)
-ifneq ($(UNTRUSTED_SYSTEM_LIBM),0)
-JLDFLAGS += $(WHOLE_ARCHIVE) $(build_libdir)/libopenlibm.a $(NO_WHOLE_ARCHIVE)
```
does julia-config handle this and the llvm-config output?

ihnorton replied:
libm yes, llvm-config no. Doesn't libjulia link llvm already?

tkelman replied:
yes but presumably the llvm-config --ldflags call was needed for something

ihnorton replied:
That was added in 072a861828378 when the file was created. My guess is copy-paste or prior/local build issue. If it breaks the example we can put it back.

tkelman commented on test/examples.jl:
```diff
@@ -89,3 +89,10 @@ if !zmq_found
+
+# test that the embedding example runs without error
+let
+    lines = readlines(pipeline(`$(joinpath(JULIA_HOME,Base.DATAROOTDIR,"doc","julia","examples","embedding","embedding"))`, stderr=DevNull))
```
wrap long line

tkelman commented on examples/embedding/Makefile:
```diff
@@ -0,0 +1,38 @@
+	$(CC) $^ -o $@ $(CFLAGS) $(DEBUGFLAGS)
+
+clean:
+	rm -f *.o *.do
```
doesn't look like this will create any object files any more

tkelman commented:
we shouldn't install binaries in the same place as documentation and source code. installing and redistributing the compiled embedding results isn't the most useful test of this, see my comments on https://github.com/JuliaLang/julia/pull/18688

twadleigh committed:
Add test/embedding.jl

ihnorton commented:
> installing and redistributing the compiled embedding results isn't the most useful test of this, see my comments on #18688

Ok:
- I put the binary in libexec in the latest update, because I agree with Yichao (and your thumbs-up) in that thread that this could be a useful test of relocatability, which I will have a follow-up PR to make possible.
- I made the test optional, as you requested in that thread, and added to appveyor/travis scripts.

tkelman commented on examples/embedding/Makefile:
```diff
@@ -0,0 +1,37 @@
+# hooks for julia build system
+ifdef BUILDROOT
+  SRCDIR := $(abspath $(dir $(lastword $(MAKEFILE_LIST))))/
+  BINDIR := $(abspath $(BUILDROOT)/usr/libexec)/
```
`build_libexecdir`

tkelman commented on examples/embedding/Makefile:
```diff
@@ -0,0 +1,39 @@
+  $(error "Please set JULIA to the path of the target julia binary!")
+endif
+ifndef BIN
+  $(error "Please set BIN to the path of the build directory!")
```
could maybe use a bit more clarification what is meant by this

ihnorton committed:
Run embedding test on appveyor and travis

tkelman commented on examples/embedding/Makefile:
```diff
@@ -0,0 +1,39 @@
+$(BIN)/embedding$(EXE): embedding.c
+	$(CC) $^ -o $@ $(CFLAGS)
+
+$(BINDIR)/embedding-debug$(EXE): embedding.c
```
should this also be BIN ?

tkelman commented on examples/embedding/Makefile:
```diff
@@ -0,0 +1,39 @@
+debug:   $(BIN)/embedding-debug$(EXE)
+
+$(BIN)/embedding$(EXE): embedding.c
+	$(CC) $^ -o $@ $(CFLAGS)
```
might want to also explicitly error if CC isn't set, or default to `cc`

vtjnash replied:
`make` already does that

ihnorton committed:
julia-config needs -std=gnu99

ihnorton committed:
Install libexec on Windows

ihnorton committed:
Teach jl_init to find itself

tkelman commented on contrib/julia-config.jl:
```diff
@@ -70,9 +70,9 @@ function cflags()
     threading_def = threadingOn() ? "-DJULIA_ENABLE_THREADING=1 " : ""
     if is_unix()
-        return """$(threading_def)-fPIC -DJULIA_INIT_DIR=\\"$arg1\\" -I$arg2"""
+        return """-std=gnu99 $(threading_def)-fPIC -DJULIA_INIT_DIR=\\"$arg1\\" -I$arg2"""
```
does the runtime need this, or the code in the embedding example?

ihnorton replied:
the runtime.

tkelman replied:
are we using gnu extensions in our headers, or which compiler versions don't default to c99 mode?

JeffBezanson commented on examples/embedding/embedding.c:
```diff
@@ -14,7 +14,7 @@ double my_c_sqrt(double x)
 int main()
 {
-    jl_init(NULL);
+    jl_init(JULIA_INIT_DIR);
```
Does this still work with `NULL` as the argument?

JeffBezanson replied:
Oh, even better, it looks like `jl_init` takes no arguments now?

JeffBezanson commented on src/jlapi.c:
```diff
@@ -43,13 +43,28 @@ JL_DLLEXPORT void jl_init_with_image(const char *julia_home_dir,
 }
 
-JL_DLLEXPORT void jl_init(const char *julia_home_dir)
+JL_DLLEXPORT void jl_init()
```
Should be `jl_init(void)`.

JeffBezanson commented on src/julia.h:
```diff
@@ -1309,12 +1309,15 @@ typedef enum {
 JL_DLLEXPORT void jl_atexit_hook(int status);
 JL_DLLEXPORT void JL_NORETURN jl_exit(int status);
+JL_DLLEXPORT const char* jl_pathname_for_handle(void *handle);
+JL_DLLEXPORT const char* jl_get_default_sysimg_path();
```
This is duplicated and needs `(void)`.

JeffBezanson commented:
Bump.

JeffBezanson commented:
Moved to #21299.
